pennylane/devices/qubit/adjoint_jacobian.py

```diff
@@ -0,0 +1,167 @@
+    sum_axes = tuple(range(1, num_wires + 1))
+    return qml.math.real(qml.math.sum(qml.math.conj(bra) * ket, axis=sum_axes))
+
+
+def adjoint_jacobian(
+    tape: QuantumTape, prep_operation: qml.operation.StatePrep = None
+):  # pylint: disable=too-many-statements
+    """Implements the adjoint method outlined in
+    `Jones and Gacon <https://arxiv.org/abs/2009.02823>`__ to differentiate an input tape.
+
+    After a forward pass, the circuit is reversed by iteratively applying adjoint
+    gates to scan backwards through the circuit.
+
+    .. note::
+        The adjoint differentiation method has the following restrictions:
+
+        * As it requires knowledge of the statevector, only statevector simulator devices can be
+          used.
+
+        * Only expectation values are supported as measurements.
+
+        * Does not work for parametrized observables like :class:`~.Hamiltonian`,
+          :class:`~.Sum` or :class:`~.Hermitian`.
+
+    Args:
+        tape (.QuantumTape): circuit that the function takes the gradient of
+        prep_operation (.StatePrep): state preparation operation applied for state initialization
+
+    Returns:
+        array or tuple[array]: the derivative of the tape with respect to trainable parameters.
+        Dimensions are ``(len(observables), len(trainable_params))``.
+
+    Raises:
+        QuantumFunctionError: if the input tape has measurements that are not expectation values
+            or contains a multi-parameter operation aside from :class:`~.Rot`
+    """
+
+    # Check validity of measurements
+    for m in tape.measurements:
+        if not isinstance(m, ExpectationMP):
+            raise qml.QuantumFunctionError(
+                "Adjoint differentiation method does not support"
+                f" measurement {m.__class__.__name__}"
+            )
+
+        if m.obs.name in {"Hamiltonian"}:
+            raise qml.QuantumFunctionError(
+                f"Adjoint differentiation method does not support observable {m.obs.name}."
+            )
+
+        if not hasattr(m.obs, "base_name"):
+            m.obs.base_name = None  # This is needed for when the observable is a tensor product
+
+    # Map wires if custom wire labels used
+    if set(tape.wires) != set(range(tape.num_wires)):
+        wire_map = {w: i for i, w in enumerate(tape.wires)}
+        tape = qml.map_wires(tape, wire_map)
+
+    # Initialization of state
+    ket = create_initial_state(
+        wires=tape.wires, prep_operation=prep_operation
+    )  #  ket(0) if prep_operation is None, else
+    for op in tape.operations:
+        if op.name not in ("QubitStateVector", "BasisState", "Snapshot"):
+            ket = apply_operation(op, ket)
+
+    n_obs = len(tape.observables)
+    bras = np.empty([n_obs] + [2] * len(tape.wires), dtype=np.complex128)
+    for kk in range(n_obs):
+        bras[kk, ...] = apply_operation(tape.observables[kk], ket)
```

Comment:
We could probably add `Hamiltonian` and `SparseHamiltonian` support to this method by registering functions to `apply_operation` that use the sparse matrix. But that would be a separate task.

Reply:
does `kk` mean something here? also nit, but you can use `for kk, obs in enumerate(tape.observables):`

Reply:
It's just an iterating index. Lots of copy pasting was involved and I forgot about `enumerate`